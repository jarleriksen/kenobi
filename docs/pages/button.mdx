---
name: Button
route: /button
---

import { Playground } from 'docz';
import styled from 'styled-components';
import Button from '../../src/button/src/Button.tsx';
import PurpleButton from '../components/PurpleButton';
import Row from '../components/Row';
import Component from '@reach/component-component';

# Button

### `Button` is a base component of Kenobi. It can contain a label and an icon before or after the label.

## Usage

<Playground>
  <Button>Base button</Button>
</Playground>

## Variations

**Component and Row are not a part of Kenobi**

<Playground>
  <Component initialState={{ isLoading: false }}>
    {({ state, setState }) => (
      <React.Fragment>
        <div style={{ marginBottom: '12px' }}>
          <Button
            onClick={() => setState({ isLoading: !state.isLoading })}
            marginBottom={4}>
            {state.isLoading ? 'Hide' : 'Show'} loading state
          </Button>
        </div>
        <Row>
          <Button marginRight={8} isLoading={state.isLoading}>
            Default
          </Button>
          <Button intent="primary" marginRight={8} isLoading={state.isLoading}>
            Primary
          </Button>
          <Button intent="success" marginRight={8} isLoading={state.isLoading}>
            Success
          </Button>
          <Button intent="warning" marginRight={8} isLoading={state.isLoading}>
            Warning
          </Button>
          <Button intent="danger" marginRight={8} isLoading={state.isLoading}>
            Danger
          </Button>
          <Button marginRight={8} iconBefore="edit" isLoading={state.isLoading}>
            Icon before
          </Button>
          <Button marginRight={8} iconAfter="edit" isLoading={state.isLoading}>
            Icon after
          </Button>
        </Row>
      </React.Fragment>
    )}
  </Component>
</Playground>

[More information on usage of Icons](/icon)

## Sizing

Default size for buttons are 32px.
`Button` will handle its own scaling based on the height.

<Playground>
  <Row>
    <Button height={24} marginRight={8}>
      Height 24
    </Button>
    <Button height={32} marginRight={8}>
      Height 32
    </Button>
    <Button height={40} marginRight={8}>
      Height 40
    </Button>
    <Button height={48} marginRight={8}>
      Height 48
    </Button>
    <Button height={56} marginRight={8}>
      Height 56
    </Button>
  </Row>
</Playground>

## Usage as a link

Due to using Styled-components as our CSS-in-JS library it is really simple to switch out the underlying
HTML element. We just have to pass the prop **isLink** to the Button component.
The button will pass the **as** prop to the Styled element.

This will render the button as an **anchor** element.

<Playground>
  <Button isLink={true} href="/button#button-as-a-link">
    I am a link!
  </Button>
</Playground>

## Extending styles

It's very easy to extend Button styles with styled-components. (Haven't tested any other solutions).

<PurpleButton>I am now Purple!</PurpleButton>

```jsx
import styled from 'styled-components';
import { Button } from 'kenobi-ui';

const PurpleButton = styled(Button)`
  background-color: #551a8b;
`;

<PurpleButton>I am now Purple!</PurpleButton>;
```
