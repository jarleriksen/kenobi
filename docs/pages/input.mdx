---
name: Input
route: /input
---

import { Playground } from 'docz';
import { Input, TextField } from 'kenobi-ui';

# Input

## Usage

Input is a wrapper around the native `input` element with Kenobi theming.
All native `input` props are accepted.

<Playground>
  <Input name="project_language" placeholder="Enter something" />
</Playground>

## TextField

`TextField` wraps `Input` but allows for passing of additional information.

<Playground>
  <TextField
    name="project_language"
    label="A required text input field"
    placeholder="Enter some text...."
    description="Text input description"
    isRequired
    errorMessage="This field is required"
  />
</Playground>
