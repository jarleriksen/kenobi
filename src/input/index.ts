export { default as Input } from './src/Input';
export { default as TextField } from './src/TextField';
export { InputIcon } from './src/InputIcon';
