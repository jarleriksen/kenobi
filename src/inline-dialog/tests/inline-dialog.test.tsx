import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { InlineDialog } from '../';
import { Button } from '../../button';
import { Input } from '../../input';
import '@testing-library/jest-dom/extend-expect';

describe('Inline Dialog', () => {
  test('should not be in document when closed', () => {
    const { getByTestId, asFragment } = render(
      <InlineDialog trigger={<div>Open</div>}>Test</InlineDialog>,
    );

    const inlineDialog = document.querySelector(
      '[data-testid="kenobi-inline-dialog"]',
    );

    expect(inlineDialog).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('should be in document when open', () => {
    const { getByTestId, asFragment, container } = render(
      <InlineDialog trigger={<Button>Open</Button>}>Test</InlineDialog>,
    );

    fireEvent.click(getByTestId('kenobi-button'));

    expect(getByTestId('kenobi-inline-dialog')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('should be in document on manual show', () => {
    const { getByTestId, asFragment, container } = render(
      <InlineDialog show={true} trigger={<Button>Open</Button>}>
        sup
      </InlineDialog>,
    );

    expect(getByTestId('kenobi-inline-dialog')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('can focus element inside dialog', () => {
    const { getByTestId, asFragment, container } = render(
      <InlineDialog trigger={<Button>Open</Button>}>
        <Input data-autofocus />
      </InlineDialog>,
    );

    fireEvent.click(getByTestId('kenobi-button'));

    expect(getByTestId('kenobi-input')).toHaveFocus();
    expect(asFragment()).toMatchSnapshot();
  });
});
