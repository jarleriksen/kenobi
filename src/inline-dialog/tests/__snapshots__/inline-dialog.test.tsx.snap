// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inline Dialog can focus element inside dialog 1`] = `
<DocumentFragment>
  <div
    class="sc-ifAKCX fMdaFI"
  >
    <div
      aria-expanded="true"
      aria-haspopup="true"
      class="sc-EHOje eQYihx"
      role="button"
    >
      <button
        class="sc-gzVnrw cJPjDm"
        data-testid="kenobi-button"
      >
        <span
          class="sc-htoDjs gPLHrW"
        >
          Open
        </span>
      </button>
    </div>
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="1"
    />
    <div
      data-focus-lock-disabled="false"
    >
      <div
        class="sc-bdVaJa ftkuoC"
        data-testid="kenobi-inline-dialog"
        id="kenobi-inline-dialog-4"
        role="dialog"
        style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
      >
        <input
          class="sc-dnqmqq cLaWQh"
          data-autofocus="true"
          data-testid="kenobi-input"
        />
      </div>
    </div>
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>
</DocumentFragment>
`;

exports[`Inline Dialog should be in document on manual show 1`] = `
<DocumentFragment>
  <div
    class="sc-ifAKCX fMdaFI"
  >
    <div
      aria-expanded="true"
      aria-haspopup="true"
      class="sc-EHOje eQYihx"
      role="button"
    >
      <button
        class="sc-gzVnrw cJPjDm"
        data-testid="kenobi-button"
      >
        <span
          class="sc-htoDjs gPLHrW"
        >
          Open
        </span>
      </button>
    </div>
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="1"
    />
    <div
      data-focus-lock-disabled="false"
    >
      <div
        class="sc-bdVaJa ftkuoC"
        data-testid="kenobi-inline-dialog"
        id="kenobi-inline-dialog-3"
        role="dialog"
        style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
      >
        <p
          class="sc-htpNat duswoF"
          color="default"
          size="400"
        >
          sup
        </p>
      </div>
    </div>
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>
</DocumentFragment>
`;

exports[`Inline Dialog should be in document when open 1`] = `
<DocumentFragment>
  <div
    class="sc-ifAKCX fMdaFI"
  >
    <div
      aria-expanded="true"
      aria-haspopup="true"
      class="sc-EHOje eQYihx"
      role="button"
    >
      <button
        class="sc-gzVnrw cJPjDm"
        data-testid="kenobi-button"
      >
        <span
          class="sc-htoDjs gPLHrW"
        >
          Open
        </span>
      </button>
    </div>
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="1"
    />
    <div
      data-focus-lock-disabled="false"
    >
      <div
        class="sc-bdVaJa ftkuoC"
        data-testid="kenobi-inline-dialog"
        id="kenobi-inline-dialog-2"
        role="dialog"
        style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
      >
        <p
          class="sc-htpNat duswoF"
          color="default"
          size="400"
        >
          Test
        </p>
      </div>
    </div>
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>
</DocumentFragment>
`;

exports[`Inline Dialog should not be in document when closed 1`] = `
<DocumentFragment>
  <div
    class="sc-ifAKCX fMdaFI"
  >
    <div
      aria-haspopup="true"
      class="sc-EHOje eQYihx"
      role="button"
    >
      <div>
        Open
      </div>
    </div>
  </div>
</DocumentFragment>
`;
